src/Counter/index.js

```diff
@@ -14,52 +16,76 @@ const { func, number, string } = PropTypes;
  */
 class Counter extends PureComponent {
   state = {
-    count: this.props.value, // eslint-disable-line react/destructuring-assignment,
+    count: this.props.defaultCount, // eslint-disable-line react/destructuring-assignment,
   };
 
   /**
-   * Component Did Update
-   * Method invoked immediately after updating occur.
+   * Returns true if requested prop exists
+   *
+   * @param {string} prop - requested property name
+   *
+   * @return {boolean}
    */
-  componentDidUpdate(prevProps) {
-    const { value } = this.props;
-
-    // Update state only if props are changed
-    if (prevProps.value !== value) {
-      this.setState({
-        count: value,
-      });
-    }
-  }
+  isControlled = prop => {
+    const props = this.props;
+
+    return props.hasOwnProperty(prop);
+  };
+
+  /**
+   * Returns the value a property from props if prop exists and from state otherwise
+   *
+   * @param {string} key - requested property name
+   *
+   * @return {*} value of the property
+   */
+  getControllableValue = key => {
+    const props = this.props;
+    const state = this.state;
+
+    return this.isControlled(key) ? props[key] : state[key];
+  };
 
   /**
    * Handle Increment click.
    */
   handleIncrementClick = () => {
-    const { count } = this.state;
+    const count = this.getControllableValue('count');
     const { step, max, onIncrement } = this.props;
 
-    this.setState(
-      prevState => ({
-        count: count === max ? prevState.count : prevState.count + step,
-      }),
-      () => (count === max ? null : onIncrement()),
-    );
+    const nextCount = Math.min(max, count + step);
+
+    if (this.isControlled('count')) {
+      count === max ? null : onIncrement(nextCount);
+    } else {
+      this.setState(
+        {
+          count: count === max ? count : nextCount,
+        },
+        () => (count === max ? null : onIncrement(nextCount)),
+      );
+    }
```

Comment:
Since the value won't be updated and noticed to the parent due to the ternary guards, I think you don't need the `Math.min`.
And regarding the legacy code, there is a mistake because it only works if step === 1 (Thanks Anto). You need to check if the potential future value does not exceed the max value so something like that:

```
handleIncrementClick = () => {
    const count = this.getControllableValue('count');
    const { step, max, onIncrement } = this.props;

    const nextCount = count + step;

    if (this.isControlled('count')) {
      nextCount > max ? null : onIncrement(nextCount);
    } else {
      this.setState(
        {
          count: nextCount > max ? count : nextCount,
        },
        () => (nextCount > max ? null : onIncrement(nextCount)),
      );
    }
```